### PR TITLE
introduced time limit per client

### PR DESCRIPTION
**Pull request by paradix**
While the feature of the time limit is nice, I thought it would be even nicer to have the option to specify a limit for each of the clients. Here is my try on that.

### CODE REVIEW

File: clientbuffer.cpp [hunk 1]
```diff
@@ -52,9 +52,14 @@ class CClientBufferMod : public CModule
     MODCONSTRUCTOR(CClientBufferMod)
     {
         AddHelpCommand();
-        AddCommand("AddClient", static_cast<CModCommand::ModCmdFunc>(&CClientBufferMod::OnAddClientCommand), "<identifier>", "Add a client.");
-        AddCommand("DelClient", static_cast<CModCommand::ModCmdFunc>(&CClientBufferMod::OnDelClientCommand), "<identifier>", "Delete a client.");
-        AddCommand("ListClients", static_cast<CModCommand::ModCmdFunc>(&CClientBufferMod::OnListClientsCommand), "", "List known clients.");
+        AddCommand("AddClient", t_d("<identifier> [timelimit]"), t_d("Add a client with optiona time limit."),
+                   [=](const CString& sLine) { OnAddClientCommand(sLine); });
```
Comment (CyberShadow) [inline clientbuffer.cpp]: Could you please rationalize the use of C++ lambdas here? As I understand this introduces a dependency on newer versions of C++ compilers.

File: clientbuffer.cpp [hunk 2]
```diff
@@ -52,9 +52,14 @@ class CClientBufferMod : public CModule
...
+        AddCommand("DelClient", t_d("<identifier>"), t_d("Delete a client."),
+                   [=](const CString& sLine) { OnDelClientCommand(sLine); });
+        AddCommand("ChangeClient", t_d("<identifier> [timelimit]"), t_d("Change a clients time limit."),
```
Comment (CyberShadow) [inline clientbuffer.cpp]: If more per-client parameters are added, they would all have to be specified all at once in the `ChangeClient` command, including the old values which the user doesn't want to change. I think it would be better if there was a `SetClientTimeLimit` command instead.

File: clientbuffer.cpp [hunk 3]
```diff
@@ -52,9 +52,14 @@ class CClientBufferMod : public CModule
     MODCONSTRUCTOR(CClientBufferMod)
     {
         AddHelpCommand();
-        AddCommand("AddClient", static_cast<CModCommand::ModCmdFunc>(&CClientBufferMod::OnAddClientCommand), "<identifier>", "Add a client.");
-        AddCommand("DelClient", static_cast<CModCommand::ModCmdFunc>(&CClientBufferMod::OnDelClientCommand), "<identifier>", "Delete a client.");
-        AddCommand("ListClients", static_cast<CModCommand::ModCmdFunc>(&CClientBufferMod::OnListClientsCommand), "", "List known clients.");
+        AddCommand("AddClient", t_d("<identifier> [timelimit]"), t_d("Add a client with optiona time limit."),
```
Comment (CyberShadow) [inline clientbuffer.cpp]: This module is not localized. Making strings go through the localization system is a pointless overhead unless you are also adding localization. Also, such changes need to be in their own commit / pull request.

Comment (CyberShadow) [reply clientbuffer.cpp]: As mentioned previously, if more per-client parameters were to be added, they would have to be all specified one after the other, without the possibility to skip any.

Specifying settings here is also superfluous to `ChangeClient` / `SetClientTimeLimit`, so I suggest not changing the semantics of the `AddClient` command.

File: clientbuffer.cpp [hunk 4]
```diff
@@ -65,8 +70,6 @@ class CClientBufferMod : public CModule
         {
             if (Args[n].Equals("autoadd", CString::CaseInsensitive))
                 m_bAutoAdd = true;
-            else if (Args[n].StartsWith("timelimit=", CString::CaseInsensitive))
-                m_iTimeLimit = Args[n].Token(1, false, "=").ToInt();
```
Comment (CyberShadow) [inline clientbuffer.cpp]: Please do not remove features or break backwards-compatibility.

File: clientbuffer.cpp [hunk 5]
```diff
@@ -556,13 +581,14 @@ void CClientBufferMod::UpdateTimestamp(const CClient* client, const CString& tar
 }
 #endif
 
-bool CClientBufferMod::WithinTimeLimit(const timeval& tv)
+bool CClientBufferMod::WithinTimeLimit(const timeval& tv, const CString& identifier)
 {
-	if (!m_iTimeLimit)
+    int timeLimit = GetNV(identifier + "/timelimit").ToInt();
```
Comment (CyberShadow) [inline clientbuffer.cpp:600]: Is there a way to achieve this without introducing a `GetNV` (lookup of string key in hash map) for every operation, even when the user never used the feature added here?

**Review by CyberShadow — CHANGES_REQUESTED** [review]
This needs some work, sorry!

File: clientbuffer.cpp (hunk 1 above)
Comment (paradix) [reply clientbuffer.cpp]: I did that because most of modules provided out of the box with znc use that kind of construction. Besides that it doesn't matter for me which way you wanna go with that.

File: clientbuffer.cpp (hunk 2 above)
Comment (paradix) [reply clientbuffer.cpp]: Yes, I have been considering that and you are right here. But it's just a command name change or you would see a different logic here?

File: clientbuffer.cpp (hunk 4 above)
Comment (paradix) [reply clientbuffer.cpp]: means you want to keep the general timelimit as is. with the possibility to change client specific timelimit?

File: clientbuffer.cpp (hunk 1 above)
Comment (CyberShadow) [reply clientbuffer.cpp]: > I did that because most of modules provided out of the box with znc

I'm guessing you're referring to the current version of ZNC. This module remains compatible with older versions, and I think that should be preserved unless a good reason comes up to break that compatibility.

File: clientbuffer.cpp (hunk 2 above)
Comment (CyberShadow) [reply clientbuffer.cpp]: Yes, in this case, it would be just a command name change.

File: clientbuffer.cpp (hunk 4 above)
Comment (CyberShadow) [reply clientbuffer.cpp]: Yes, I think it would be good to leave the old setting as the default if a per-client one is not specified.

File: clientbuffer.cpp (hunk 5 above)
Comment (paradix) [reply clientbuffer.cpp:600]: The other things are easy fixed, but this one gives me headaches :)
Can't think for now of a solution that wouldn't require storing the client timelimit somewhere and retrieving it by the identifier key. 
Do you have any thoughts on that?

Comment (paradix) [reply clientbuffer.cpp:600]: You could create some kind of an object representing the client in the OnClientLogin Method.
But since you can have many clients connecting simultaneously you would than still have to find that object referring to the given client identifier. 
So the question here is if this would be much more efficient than using the std::map stored in memory that is used as the registry?

File: clientbuffer.cpp [hunk 6]
```diff
@@ -52,9 +52,10 @@ class CClientBufferMod : public CModule
     MODCONSTRUCTOR(CClientBufferMod)
     {
         AddHelpCommand();
-        AddCommand("AddClient", static_cast<CModCommand::ModCmdFunc>(&CClientBufferMod::OnAddClientCommand), "<identifier>", "Add a client.");
+        AddCommand("AddClient", static_cast<CModCommand::ModCmdFunc>(&CClientBufferMod::OnAddClientCommand), "<identifier> [timelimit]", "Add a client with optional time limit.");
```
Comment (CyberShadow) [inline clientbuffer.cpp]: Is there any reason to have two ways to specify the time limit (`AddClient` and `SetClientTimeLimit`)?

Specifying client settings in `AddClient` has the same design issue as `ChangeClient`.

Sometimes, specifying options when creating some object is necessary for security reasons (e.g. `open(fn, 0600)` instead of (`open(fn) + chmod(fn, 0600)`), but I don't see the need here.

File: clientbuffer.cpp [hunk 7]
```diff
@@ -52,9 +52,10 @@ class CClientBufferMod : public CModule
...
         AddCommand("DelClient", static_cast<CModCommand::ModCmdFunc>(&CClientBufferMod::OnDelClientCommand), "<identifier>", "Delete a client.");
         AddCommand("ListClients", static_cast<CModCommand::ModCmdFunc>(&CClientBufferMod::OnListClientsCommand), "", "List known clients.");
+        AddCommand("SetClientTimeLimit", static_cast<CModCommand::ModCmdFunc>(&CClientBufferMod::OnSetClientTimeLimit), "<identifier> [timelimit]", "Change a clients time limit.");
```
Comment (CyberShadow) [inline clientbuffer.cpp]: - I think it doesn't make sense for `timelimit` to be optional here, so it should be specified as `<timelimit>` in the command description.
- Typo (clients -> client's)

```suggestion
        AddCommand("SetClientTimeLimit", static_cast<CModCommand::ModCmdFunc>(&CClientBufferMod::OnSetClientTimeLimit), "<identifier> <timelimit>", "Change a client's time limit.");           
```

File: clientbuffer.cpp [hunk 8]
```diff
@@ -162,13 +167,32 @@ void CClientBufferMod::OnDelClientCommand(const CString& line)
...
 }
 
+/// Callback for the SetClientTimeLimit module command.
+void CClientBufferMod::OnSetClientTimeLimit(const CString& line)
+{
+    const CString identifier = line.Token(1);
+    const int timeLimit = line.Token(2).ToInt();
+
+    if (identifier.empty()) {
+        PutModule("Usage: ChangeClient <identifier> [timelimit]");
```
Comment (CyberShadow) [inline clientbuffer.cpp]: As above.
```suggestion
        PutModule("Usage: ChangeClient <identifier> <timelimit>");
```

File: clientbuffer.cpp [hunk 9]
```diff
@@ -134,15 +136,18 @@ class CClientBufferMod : public CModule
...
+        PutModule("Usage: AddClient <identifier> [timelimit]");
         return;
     }
     if (HasClient(identifier)) {
         PutModule("Client already exists: " + identifier);
         return;
     }
-    AddClient(identifier);
+
+    AddClient(identifier, timeLimit > 0 ? timeLimit : m_iTimeLimit);
```
Comment (CyberShadow) [inline clientbuffer.cpp]: This doesn't look right. If the user adds a client without specifying a time limit, and then changes the global time limit, the change will not be applied to the client.

I think it would be more correct to make `WithinTimeLimit` use the per-client time limit if it has been set by the user (is non-zero), and if not, use the global time limit.

File: clientbuffer.cpp (hunk 5 above)
Comment (CyberShadow) [reply clientbuffer.cpp:600]: I think your analysis is correct. The only improvement over this that I can see would be to use the `Client*` as a map key directly, which would save us from hashing the string every time.

Let's leave this alone for now. We can revisit it later if needed.

File: clientbuffer.cpp [hunk 10]
```diff
@@ -134,14 +137,16 @@ class CClientBufferMod : public CModule
 void CClientBufferMod::OnAddClientCommand(const CString& line)
 {
     const CString identifier = line.Token(1);
+
     if (identifier.empty()) {
-        PutModule("Usage: AddClient <identifier>");
+        PutModule("Usage: AddClient <identifier> <timelimit>");
```
Comment (CyberShadow) [inline clientbuffer.cpp]: Usage string still has the timelimit parameter:

```suggestion
        PutModule("Usage: AddClient <identifier>");
```

File: clientbuffer.cpp [hunk 11]
```diff
@@ -162,13 +167,32 @@ void CClientBufferMod::OnDelClientCommand(const CString& line)
     PutModule("Client removed: " + identifier);
 }
 
+/// Callback for the SetClientTimeLimit module command.
+void CClientBufferMod::OnSetClientTimeLimit(const CString& line)
+{
+    const CString identifier = line.Token(1);
+    const int timeLimit = line.Token(2).ToInt();
+
+    if (identifier.empty() || !timeLimit) {
```
Comment (CyberShadow) [inline clientbuffer.cpp]: If we reject a per-client time limit of zero, how would the user clear the per-client time limit so that the module uses the global value again?

Maybe, instead:
```diff
diff --git a/clientbuffer.cpp b/clientbuffer.cpp
index 3ecd4aa..2733389 100644
--- a/clientbuffer.cpp
+++ b/clientbuffer.cpp
@@ -55,7 +55,7 @@ public:
         AddCommand("AddClient", static_cast<CModCommand::ModCmdFunc>(&CClientBufferMod::OnAddClientCommand), "<identifier>", "Add a client.");
         AddCommand("DelClient", static_cast<CModCommand::ModCmdFunc>(&CClientBufferMod::OnDelClientCommand), "<identifier>", "Delete a client.");
         AddCommand("ListClients", static_cast<CModCommand::ModCmdFunc>(&CClientBufferMod::OnListClientsCommand), "", "List known clients.");
-        AddCommand("SetClientTimeLimit", static_cast<CModCommand::ModCmdFunc>(&CClientBufferMod::OnSetClientTimeLimit), "<identifier> <timelimit>", "Change a client's time limit.");           
+        AddCommand("SetClientTimeLimit", static_cast<CModCommand::ModCmdFunc>(&CClientBufferMod::OnSetClientTimeLimit), "<identifier> [timelimit]", "Change a client's time limit.");           
         AddTimer(new CClientBufferCacheJob(this, 1 /* sec */, 0, "ClientBufferCache", "Periodically save ClientBuffer registry to disk"));
     }
 
@@ -173,8 +173,8 @@ void CClientBufferMod::OnSetClientTimeLimit(const CString& line)
     const CString identifier = line.Token(1);
     const int timeLimit = line.Token(2).ToInt();
 
-    if (identifier.empty() || !timeLimit) {
-        PutModule("Usage: SetClientTimeLimit <identifier> <timelimit>");
+    if (identifier.empty()) {
+        PutModule("Usage: SetClientTimeLimit <identifier> [timelimit]");
         return;
     }
     if (!HasClient(identifier)) {
@@ -182,7 +182,10 @@ void CClientBufferMod::OnSetClientTimeLimit(const CString& line)
         return;
     }
     SetClientTimeLimit(identifier, timeLimit);
-    PutModule("Client's " + identifier +  " changed time limit: " + CString(timeLimit) );
+    if (timeLimit)
+        PutModule("Client's " + identifier +  " changed time limit: " + CString(timeLimit) );
+    else
+        PutModule("Client's " + identifier +  " cleared time limit.");
 }
 
 /// Callback for the ListClients module command.
@@ -469,7 +472,10 @@ bool CClientBufferMod::HasClient(const CString& identifier)
 bool CClientBufferMod::SetClientTimeLimit(const CString& identifier, const int timeLimit)
 {
     m_bDirty = true;
-    return SetNV(identifier + "/timelimit", CString(timeLimit), false);
+    if (timeLimit)
+        return SetNV(identifier + "/timelimit", CString(timeLimit), false);
+    else
+        return DelNV(identifier + "/timelimit", false);
 }
 
 #if ZNC17
```

**Review by CyberShadow — COMMENTED** [review]
Almost there :)

File: clientbuffer.cpp (hunk 11 above)
Comment (paradix) [reply clientbuffer.cpp]: Yeah, seems like a good point.

**Review by CyberShadow — APPROVED** [review]
Thank you!
